### PR TITLE
feat(protocol-designer): add functionality to multi select toolbar

### LINKED ISSUE #7105: PD: multi select controls + functionality Pt 1 (includes exit batch mode)
## User story
As a user I want the multi select controls to be functional, and have a way to exit batch edit mode

**Please note** This ticket excludes the 'Delete' and 'Duplicate' functionality. That ticket is https://app.zenhub.com/workspaces/opentrons-spddrs-5e1f591ca60369aa0e6b8eaf/issues/opentrons/opentrons/7127

## Acceptance criteria for control bar at the top of the timeline
[This figma chart](https://www.figma.com/file/6p1uxeeioDikcfSwLCfxi4/Batch-Editing?node-id=705%3A16693) has the below information in a visual manner 
**Select/de-select all**
- [ ] There is an icon with two states that represents whether steps in the timeline are selected
- [ ] This icon has the tooltip "Select All". If all are selected, tooltip should read "Deselect All"
- [ ] When there are some checked steps in the timeline and some unchecked steps, the icon is in the _mixed_ state 
- [ ] When icon is in _mixed_ state and it is clicked, two things happen: (1) the icon switches to _all selected_ state (2) all non-terminal steps in the timeline are selected
- [ ] When all non-terminal steps in the timeline are selected the icon is in the _all selected_ state
- [ ] When icon is in _all selected_ state and is clicked PD attempts to exit multi select mode. When exiting multi select mode, make sure that single edit mode is entered with the `lastSelected` step being selected. Please see [figma chart ](https://www.figma.com/file/6p1uxeeioDikcfSwLCfxi4/Batch-Editing?node-id=827%3A67)for conditions as to what happens next. 

**Expand/collapse**
- [ ] There is an icon/button for expand/collapse
- [ ] When multi-select mode is first entered the icon is in the _expand_ state
- [ ] When icon is in the _expand_ state it has a tooltip that reads "Expand"
- [ ] When icon is in the _expand_ state and is clicked two things happen: (1) the icon switches to the _collapse_ state (2) all selected steps are expanded
- [ ] When icon is in the _collapse_ state it has a tooltip that reads "Collapse"
- [ ] When icon is in the _collapse_ state and is clicked two things happen: (1) the icon switches to the _expand_ state (2) all selected steps are collapsed

**Test Steps for control bar:**

1. Create a simple protocol with 4 transfer steps 
2. Enter batch edit mode by selecting transfer steps 2&3

Expected Result: 
- Control bar appears at top of the timeline only after batch edit is entered. 
- Selected steps icon is in the mixed state at the left hand side of the control bar. 
- The right side of the control bar displays delete, duplicate, and expand icon 
- Steps 2&3 in timeline are checked.

3. Click on selected steps icon to select all non terminal steps. 

Expected Result: 
- All non terminal steps should now be selected, and appear with a check next to them in the timeline. 
- Selected steps icon changes to all steps selected mode.

4. Hover over the expand icon until the tooltip displays, then click on the expand icon.

Expected Result:
- Tooltip appears upon hover, tooltip reads "Expand"
- All transfer steps should now be expanded in the timeline.
- Expand icon has changed to collapse icon

5. Deselect step 1
6. Hover over the collapse icon until the tooltip displays, then click on the collapse icon

Expected Result:
- Step 1 no longer has a check next to it, and the step is still expanded in the timeline
- Steps 2-4 are still checked, and are now collapsed in the timeline
- Collapse icons has changed back to expand icon
- Selected steps icon has changed back to mixed state (dash)

7. Select step 1

Expected Result:
- Selected steps icon has changed to all non-terminal steps selected (check)

8. Click on all steps selected icon.

Expected Result:
- All steps are deselected aside from the first step in the selection
- Batch edit mode is exited and PD goes back to single edit mode
- Control bar in the timeline goes away

### LINKED ISSUE #7127: PD: multi select controls + functionality Pt 2 (Delete & Duplicate)
**User story**
As a user I want to delete and duplicate multiple steps

## Acceptance criteria for control bar at the top of the timeline
[This figma chart](https://www.figma.com/file/6p1uxeeioDikcfSwLCfxi4/Batch-Editing?node-id=705%3A16693) has the below information in a visual manner 
**Delete**
- [ ] There is an icon for deleting steps
- [ ] This icon has the tooltip "Delete"
- [ ] After step deletion, PD will remain in batch edit mode and have only 1 step selected
- [ ] After the deletion of steps, the next step selected is one of two cases: (1) If the step following the last deleted step is a **non terminal** item, that step is the next step selected. (2) If the step following the last deleted step is a **terminal item**, select the last non-terminal step in the protocol instead.

If all steps get deleted, exit batch edit mode with starting deck state selected

**Duplicate**
- [ ] There is an icon for duplicating steps
- [ ] This icon has the tooltip "Duplicate"
- [ ] Steps created by the duplication are contiguous, even if the initial selection wasn't
- [ ] Duplicated steps appear below the last step of the original selection.
- [ ] Only the duplicates are now selected. The original selection is not selected.

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview

This PR adds functionality to the multi select toolbar. Closes #7105 Closes #7127

# Changelog

- Add multi step deletion functionality to multi select toolbar
- Add multi step collapse/expand functionality to multi select toolbar
- Add multi step select/deselect functionality to multi select toolbar
- Add multi step duplication functionality to multi select toolbar

# Review requests

Code review, including tests. There's a lot here plz make sure it's readable.

Enter batch edit mode, and follow the test cases in [the ticket](https://github.com/Opentrons/opentrons/issues/7105 ). Make sure you can properly:

(1) select/deselect multiple steps
(2) delete multiple steps
(3) duplicate multiple steps
(4) expand/collapse multiple steps

# Risk assessment
Low/med, not being used yet and pretty well tested.

### CODE REVIEW

File: protocol-designer/src/ui/steps/test/reducers.test.js [hunk 1]
```diff
@@ -71,6 +71,24 @@ describe('collapsedSteps reducer', () => {
       '4': true,
     })
   })
+  it('should toggle multiple steps upon TOGGLE_MULTIPLE_STEPS_COLLAPSED', () => {
```
Comment (Kadee80) [inline protocol-designer/src/ui/steps/test/reducers.test.js]: Isn't the desired behavior an expand all / collapse all selected, not toggle the expanded vs collapsed ones?

Comment (shlokamin) [reply protocol-designer/src/ui/steps/test/reducers.test.js]: It is indeed! I made an action that takes in the steps to toggle as input to:
(1) be more functional and explicit
(2) not tie concerns up too much with the UI. Ex, if in the future we need to be able to toggle some but not all steps, we should be able to. 

It's kinda like right now the `selectAllSteps` action creator calls `SELECT_MULTIPLE_STEPS` under the hood. 
if it makes more sense to make a `toggleAllStepsCollapsed` action creator that calls `TOGGLE_MULTIPLE_STEPS_COLLAPSED` under the hood, instead of a `toggleMultipleStepsCollapsed` action creator that takes in a list of ids I'm also down for that.

File: protocol-designer/src/step-forms/selectors/index.js [hunk 2]
```diff
@@ -421,6 +421,8 @@ export const getCurrentFormHasUnsavedChanges: Selector<boolean> = createSelector
   }
 )
 
+export const getBatchEditFormHasUnsavedChanges = (): boolean => false
+
```
Comment (shlokamin) [inline protocol-designer/src/step-forms/selectors/index.js:422]: this selector is captured here https://github.com/Opentrons/opentrons/issues/7245

File: protocol-designer/src/ui/steps/test/reducers.test.js (hunk 1 above)
Comment (Kadee80) [reply protocol-designer/src/ui/steps/test/reducers.test.js]: Its seems like the behavior in the UI right now toggles the open ones to closed, and the closed to open, which might not necessarily jive with the icon/tooltip in the toolbars expected behavior?

File: protocol-designer/src/steplist/actions/actions.js [hunk 3]
```diff
@@ -36,6 +42,50 @@ export const deleteStep = (stepId: StepIdType): DeleteStepAction => ({
...
+      return Math.max(currentStepIndex, highestIndex)
+    }, 0)
+    let nextStepId = orderedStepIds[highestDeletedIndex + 1]
+    while (!nextStepId) {
+      highestDeletedIndex -= 1
+      const potentialNextStepId = orderedStepIds[highestDeletedIndex]
+      if (stepIds.includes(potentialNextStepId)) {
+        continue
+      }
+      nextStepId = potentialNextStepId
```
Comment (shlokamin) [inline protocol-designer/src/steplist/actions/actions.js]: this whole block is rather unfortunate and hard to follow. i would love to refactor this to be more readable

File: protocol-designer/src/ui/steps/test/reducers.test.js (hunk 1 above)
Comment (shlokamin) [reply protocol-designer/src/ui/steps/test/reducers.test.js]: Ohhhh i see what you are saying, good catch you are absolutely correct! Will fix this now, thanks for catching this

**Review by IanLondon — CHANGES_REQUESTED** [review]
**Select/de-select all**
- [x] There is an icon with two states that represents whether steps in the timeline are selected
- [x] This icon has the tooltip "Select All". If all are selected, tooltip should read "Deselect All"
- [x] When there are some checked steps in the timeline and some unchecked steps, the icon is in the _mixed_ state 
- [x] When icon is in _mixed_ state and it is clicked, two things happen: (1) the icon switches to _all selected_ state (2) all non-terminal steps in the timeline are selected
- [x] When all non-terminal steps in the timeline are selected the icon is in the _all selected_ state
- [ ] **(??? I'm not sure what this means. What is the lastSelected supposed to become when you do "Select All" ???** When icon is in _all selected_ state and is clicked PD attempts to exit multi select mode. When exiting multi select mode, make sure that single edit mode is entered with the `lastSelected` step being selected. Please see [figma chart ](https://www.figma.com/file/6p1uxeeioDikcfSwLCfxi4/Batch-Editing?node-id=827%3A67)for conditions as to what happens next. 

**Expand/collapse**
- [x] There is an icon/button for expand/collapse
- [x] When multi-select mode is first entered the icon is in the _expand_ state
- [x] When icon is in the _expand_ state it has a tooltip that reads "Expand"
- [x] When icon is in the _expand_ state and is clicked two things happen: (1) the icon switches to the _collapse_ state (2) all selected steps are expanded
- [x] When icon is in the _collapse_ state it has a tooltip that reads "Collapse"
- [x] Seems they just toggle, so some may collapse and others may expand** When icon is in the _collapse_ state and is clicked two things happen: (1) the icon switches to the _expand_ state (2) all selected steps are collapsed

**Delete**
- [x] There is an icon for deleting steps
- [x] This icon has the tooltip "Delete"
- [x] After step deletion, PD will remain in batch edit mode and have only 1 step selected
- [x] After the deletion of steps, the next step selected is one of two cases: (1) If the step following the last deleted step is a **non terminal** item, that step is the next step selected. (2) If the step following the last deleted step is a **terminal item**, select the last non-terminal step in the protocol instead.

If all steps get deleted, exit batch edit mode with starting deck state selected

**Duplicate**
- [x] There is an icon for duplicating steps
- [x] This icon has the tooltip "Duplicate"
- [x] Steps created by the duplication are contiguous, even if the initial selection wasn't
- [x] Duplicated steps appear below the last step of the original selection.
- [x] Only the duplicates are now selected. The original selection is not selected.
- [x] ordering of duplicates ~seems unexpected. Should the selection order be preserved in timeline order, rather than in selection order?~ is OK

**Review by Kadee80 — APPROVED** [review]
💪 Thanks for fixing the toggle behavior and the tooltips!

File: protocol-designer/src/steplist/actions/actions.js [hunk 4]
```diff
@@ -36,6 +42,50 @@ export const deleteStep = (stepId: StepIdType): DeleteStepAction => ({
...
 
+export type DeleteMultipleStepsAction = {|
+  type: 'DELETE_MULTIPLE_STEPS',
+  payload: Array<StepIdType>,
+|}
+
+// delete the steps, then update the selection
+export const deleteMultipleSteps = (
+  stepIds: Array<StepIdType>
+): ThunkAction<*> => (dispatch: ThunkDispatch<*>, getState: GetState) => {
```
Comment (IanLondon) [inline protocol-designer/src/steplist/actions/actions.js]: nitpick: I think type inference by `*` is deprecated/discouraged(???), uhh cannot confirm, I can't find it in the docs. But you should be able to type this fully (eg `renameLabware` thunk has `ThunkAction<CreateContainerAction | RenameLabwareAction>`). Your call though, we do a bunch of `FunkyReduxType<*>` in PD and when we go to TS the landscape might be different

**Review by IanLondon — CHANGES_REQUESTED** [review]
everything looks good other than the changes we talked about this afternoon, LMK when to re-review

**Review by IanLondon — APPROVED** [review]
new commits look good!
